### DISCUSSION TITLE
syntax discussed with Aadit

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#425] [General]
@aaditmshah suggested a latex inspired syntax:

```latex
\article{3 Reasons Why You Should Create Your Own ESLint Config}
~tags{
\tag{eslint}
\tag{eslint-config}
\tag{ivory}
}
~content{
I recently created my own eslint config called 
[Ivory](https://www.npmjs.com/package/eslint-config-ivory). It's my first open-source 
project. I decided that since I'm going to be a full-time open-source developer, I need 
to adhere to a single code style for all my repositories. So, why didn't I go with a popular 
eslint config like [Airbnb](https://github.com/airbnb/javascript)? Here are 3 reasons why 
I created my own eslint config, and why you should consider doing the same.
}

\link{Ivory}~href{https://www.npmjs.com/package/eslint-config-ivory}
```

**Comment by aaditmshah** [comment]
Note that this markup format is only for content. It shouldn't be used for presentation, document structure, or application logic. For presentation we can use CSS, for document structure we can use HTML, and for application logic we can use JavaScript. The idea is to separate the content from the way it's used. This is something that markdown doesn't do. Markdown mixes the content with the presentation and the document structure. For example, consider [my blog post written in markdown](https://aadit.codes/blog/why-create-your-own-eslint-config).

```markdown
---
title: "3 Reasons Why You Should Create Your Own ESLint Config"
tags:
  - eslint
  - eslint-config
  - ivory
---

I recently created my own eslint config called [Ivory](https://www.npmjs.com/package/eslint-config-ivory). It's my first open-source project. I decided that since I'm going to be a full-time open-source developer, I need to adhere to a single code style for all my repositories. So, why didn't I go with a popular eslint config like [Airbnb](https://github.com/airbnb/javascript)? Here are 3 reasons why I created my own eslint config, and why you should consider doing the same.

## 1. You Will Have Full Control Over Your Code Style

I'm a very opinionated software developer. For example, I prefer writing `for…of` loops over `.forEach`. However, the Airbnb eslint config disallows `for…of` loops. So, what's a developer supposed to do? Either you give in and use `.forEach`, or you explicitly override the rule.

Unfortunately, overriding a rule is not always as simple as turning it `"off"`. For example, the Airbnb eslint config uses the [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule to disallow `for…of` loops. If you turn it `"off"` then you will inadvertently allow `for…in` loops, labels, and the `with` statement. Those are all things I want to disallow. Tweaking the rule to only allow `for…of` loops is complicated.

Now, though it might be complicated to tailor the Airbnb eslint config to suit your code style, it's nevertheless possible. But customized eslint configs suffer from another big problem — updates. Every time Airbnb updates its eslint config, there's a possibility that the updates might conflict with your code style. However, if you don't update then you skip out on configuring new rules.

All these problems can be avoided if you create your own eslint config. Since you have full control of the eslint config, there's no possibility of subverting your code style. Furthermore, you can set up automated tests and workflows to ensure that your eslint config is always consistent and up to date. For example, the [Alloy](https://github.com/AlloyTeam/eslint-config-alloy#philosophy) eslint config runs automated weekly checks for new rules added by eslint and its plugins. It also checks whether its rules conflict with [Prettier](https://prettier.io/) or are deprecated.

Sounds like a lot of work? It can indeed be daunting to create your own eslint config and manually configure all the rules to suit your code style. For example, I spent an entire 40-hour work week configuring all the rules in Ivory. But unlike me, you don't have to pay the cost upfront. You can create a fork of Ivory, start using it, and configure it over time to suit your code style.

## 2. You Will Have Consistency Across All Repositories

Creating your own eslint config does come at the significant cost of manually configuring all the rules. However, the benefit is that you only have to write the config once for all your repositories. For every new repository, you should only have to extend your eslint config. You should never have to override any rules, as that would make your code style inconsistent across repositories.

I know from experience that it's very easy for developers to create multiple repositories with inconsistent code styles. It's effortless for developers to override rules on a per-repository basis. Especially when you're working in a large organization with multiple teams. Using an external eslint config like Airbnb just encourages developers to override rules on a per-repository basis.

Now, inconsistent code styles would not be a major issue if eslint rules only enforced a stylistic preference, like `for…of` loops vs `.forEach`. However, a lot of eslint rules enforce a particular code style to prevent potential bugs. For example, I have a strong preference to ban type assertions in TypeScript by enabling the [`consistent-type-assertions`](https://typescript-eslint.io/rules/consistent-type-assertions) rule and setting the `assertionStyle` to `"never"`. This prevents a lot of bugs by forcing developers to fix errors reported by the type checker. If this code style was inconsistent across repositories then bugs may slip through.

When you create your own eslint config, you should discourage developers from overriding rules on a per-repository basis. Instead, you should encourage developers to raise an issue in your eslint config repository. This allows you to debate the proposed code style, update the eslint config in a disciplined manner, and most importantly maintain consistency across repositories.

## 3. You Can Have Customization With Preset Configs

While maintaining consistency across repositories is important, not all repositories require the same configuration rules. For example, you may only want to activate React rules for front-end applications. Fortunately, creating your own eslint config doesn't mean that you have to give up customization across repositories. You can have granular configs and combine them as required.

Breaking big things into smaller things and then recombining them to form the original thing is a common theme in computer science. Programmers know this concept by many names such as [dynamic programming](https://en.wikipedia.org/wiki/Dynamic_programming), the [combinator pattern](https://wiki.haskell.org/Combinator_pattern), and [decomposition](http://worrydream.com/LearnableProgramming/) and [recomposition](https://www.cs.kent.ac.uk/people/staff/dat/miranda/whyfp90.pdf). This technique has become popular in modern times. For example, the core concept of [Tailwind CSS](https://tailwindcss.com/), the most popular CSS framework, is combining simple style classes to build more complex styles.

The same concept can be applied to eslint configs. We already have a way of combining multiple eslint configs using the `extends` property. So, instead of creating a single monolithic eslint config, we can create multiple eslint configs at the level of granularity we desire. We can then combine them using the `extends` property into several preset configs for different kinds of repositories. The [Canonical](https://github.com/gajus/eslint-config-canonical#example-configuration) eslint config has some excellent examples on how to combine multiple configs.

## So Go Ahead and Create Your Own ESLint Config

Using a popular eslint config like Airbnb is easy when you want to get started as soon as possible. However, the return on investment sharply declines when you start overriding rules, copying your eslintrc file, and trying to maintain consistency across repositories. Having your own eslint config solves all these problems. And you don't need to start from scratch. Just fork [Ivory](https://www.npmjs.com/package/eslint-config-ivory) and get going.
```

In the above example, everything has some semantic meaning.

```markdown
Using `#` creates various heading levels.

# Heading 1
## Heading 2
### Heading 3
#### Heading 4
##### Heading 5
###### Heading 6

Using [link](href) creates a hyperlink.

Newlines between lines creates paragraphs.
```

This is problematic for two main reasons.

1. **Markdown is not extensible.** Markdown cannot markup all the content that you want. To work around this limitation, you have to extend markdown in non-standard ways. In the above example, I had to use YAML to create the frontmatter, which contains the title and tags, for my blog post. This is a big disadvantage of markdown over other markup languages like XML, which are extensible.
2. **Markdown is opinionated.** Markdown does more than just markup content. It also imposes its on semantics onto that content. For example, headings in markdown are converted into `<h1>` to `<h6>` tags, links are converted into `<a href>` tags, and paragraphs are converted into `<p>` tags. Furthermore, this semantics is enforced by the ecosystem of markdown tools. This makes it very difficult for people to use the contents of the markdown file the way they want to. Hence, you often have to use complex extensions to do things like generate a table of contents or generate a summary of a markdown document.

So, why not use XML instead of markdown? After all XML is both extensible and non-opinionated. Well, XML has its own set of problems.

1. **XML is verbose.** No big surprise here. XML documents have a very low signal-to-noise ratio. This makes it a pain to read and write XML documents. This is also one of the reasons why markdown is so popular. Markdown has a very high signal-to-noise ratio. It's very easy to read and write markdown files.
2. **XML attributes cannot be complex.** XML tags must have simple string attributes. You can't have other XML tags as attributes. This means that complex attributes need to be modeled as child nodes. This is not ideal because we would like to separate metadata from content as much as possible.

So instead of using markdown or XML, I suggest creating a new markup language which looks a little like LaTeX. Consider my above blog post written in such a language.

```
\article{3 Reasons Why You Should Create Your Own ESLint Config}~tag{eslint}~tag{eslint-config}~tag{ivory}~content{
  \paragraph{I recently created my own eslint config called \link{Ivory}~href{https://www.npmjs.com/package/eslint-config-ivory}. It's my first open-source project. I decided that since I'm going to be a full-time open-source developer, I need to adhere to a single code style for all my repositories. So, why didn't I go with a popular eslint config like \link{Airbnb}~href{https://github.com/airbnb/javascript}? Here are 3 reasons why I created my own eslint config, and why you should consider doing the same.}
  \section{1. You Will Have Full Control Over Your Code Style}~content{
    \paragraph{I'm a very opinionated software developer. For example, I prefer writing \code{for…of} loops over \code{.forEach}. However, the Airbnb eslint config disallows \code{for…of} loops. So, what's a developer supposed to do? Either you give in and use \code{.forEach}, or you explicitly override the rule.}
    \paragraph{Unfortunately, overriding a rule is not always as simple as turning it \code{"off"}. For example, the Airbnb eslint config uses the \link{\code{no-restricted-syntax}}~href{https://eslint.org/docs/latest/rules/no-restricted-syntax} rule to disallow \code{for…of} loops. If you turn it \code{"off"} then you will inadvertently allow \code{for…in} loops, labels, and the \code{with} statement. Those are all things I want to disallow. Tweaking the rule to only allow \code{for…of} loops is complicated.}
    \paragraph{Now, though it might be complicated to tailor the Airbnb eslint config to suit your code style, it's nevertheless possible. But customized eslint configs suffer from another big problem — updates. Every time Airbnb updates its eslint config, there's a possibility that the updates might conflict with your code style. However, if you don't update then you skip out on configuring new rules.}
    \paragraph{All these problems can be avoided if you create your own eslint config. Since you have full control of the eslint config, there's no possibility of subverting your code style. Furthermore, you can set up automated tests and workflows to ensure that your eslint config is always consistent and up to date. For example, the \link{Alloy}~href{https://github.com/AlloyTeam/eslint-config-alloy#philosophy} eslint config runs automated weekly checks for new rules added by eslint and its plugins. It also checks whether its rules conflict with \link{Prettier}~href{https://prettier.io/} or are deprecated.}
    \paragraph{Sounds like a lot of work? It can indeed be daunting to create your own eslint config and manually configure all the rules to suit your code style. For example, I spent an entire 40-hour work week configuring all the rules in Ivory. But unlike me, you don't have to pay the cost upfront. You can create a fork of Ivory, start using it, and configure it over time to suit your code style.}
  }
  \section{2. You Will Have Consistency Across All Repositories}~content{
    \paragraph{Creating your own eslint config does come at the significant cost of manually configuring all the rules. However, the benefit is that you only have to write the config once for all your repositories. For every new repository, you should only have to extend your eslint config. You should never have to override any rules, as that would make your code style inconsistent across repositories.}
    \paragraph{I know from experience that it's very easy for developers to create multiple repositories with inconsistent code styles. It's effortless for developers to override rules on a per-repository basis. Especially when you're working in a large organization with multiple teams. Using an external eslint config like Airbnb just encourages developers to override rules on a per-repository basis.}
    \paragraph{Now, inconsistent code styles would not be a major issue if eslint rules only enforced a stylistic preference, like \code{for…of} loops vs \code{.forEach}. However, a lot of eslint rules enforce a particular code style to prevent potential bugs. For example, I have a strong preference to ban type assertions in TypeScript by enabling the \link{\code{consistent-type-assertions}}~href{https://typescript-eslint.io/rules/consistent-type-assertions} rule and setting the \code{assertionStyle} to \code{"never"}. This prevents a lot of bugs by forcing developers to fix errors reported by the type checker. If this code style was inconsistent across repositories then bugs may slip through.}
    \paragraph{When you create your own eslint config, you should discourage developers from overriding rules on a per-repository basis. Instead, you should encourage developers to raise an issue in your eslint config repository. This allows you to debate the proposed code style, update the eslint config in a disciplined manner, and most importantly maintain consistency across repositories.}
  }
  \section{3. You Can Have Customization With Preset Configs}~content{
    \paragraph{While maintaining consistency across repositories is important, not all repositories require the same configuration rules. For example, you may only want to activate React rules for front-end applications. Fortunately, creating your own eslint config doesn't mean that you have to give up customization across repositories. You can have granular configs and combine them as required.}
    \paragraph{Breaking big things into smaller things and then recombining them to form the original thing is a common theme in computer science. Programmers know this concept by many names such as \link{dynamic programming}~href{https://en.wikipedia.org/wiki/Dynamic_programming}, the \link{combinator pattern}~href{https://wiki.haskell.org/Combinator_pattern}, and \link{decomposition}~href{http://worrydream.com/LearnableProgramming/} and \link{recomposition}~href{https://www.cs.kent.ac.uk/people/staff/dat/miranda/whyfp90.pdf}. This technique has become popular in modern times. For example, the core concept of \link{Tailwind CSS}~href{https://tailwindcss.com/}, the most popular CSS framework, is combining simple style classes to build more complex styles.}
    \paragraph{The same concept can be applied to eslint configs. We already have a way of combining multiple eslint configs using the \code{extends} property. So, instead of creating a single monolithic eslint config, we can create multiple eslint configs at the level of granularity we desire. We can then combine them using the \code{extends} property into several preset configs for different kinds of repositories. The \link{Canonical}~href{https://github.com/gajus/eslint-config-canonical#example-configuration} eslint config has some excellent examples on how to combine multiple configs.}
  }
  \section{So Go Ahead and Create Your Own ESLint Config}~content{
    \paragraph{Using a popular eslint config like Airbnb is easy when you want to get started as soon as possible. However, the return on investment sharply declines when you start overriding rules, copying your eslintrc file, and trying to maintain consistency across repositories. Having your own eslint config solves all these problems. And you don't need to start from scratch. Just fork \link{Ivory}~href{https://www.npmjs.com/package/eslint-config-ivory} and get going.}
  }
}
```

In the above example, we're marking up content using the syntax `\tag-name{content}`. You can use any tag name you want. Unlike HTML, you are not limited to a predefined set of tags. Hence, the markup language is extensible and unopinionated like XML.

Attributes can be added to the previous tag by using a tilde, i.e. `~`, instead of a backslash, i.e. `\`. For example, `\link{Aadit Codes}~href{https://aadit.codes/}` represents a link to my website, [Aadit Codes](https://aadit.codes/). You can add as many attributes as you want. You can even repeat attribute names. In the above example, I repeated the `tag` attribute name to denote a list of tags for my blog post, i.e. `~tag{eslint}~tag{eslint-config}~tag{ivory}`.

Unlike XML, attributes can have any kind of content including nested tags. It's also not as verbose as XML, and although it's not as terse as markdown yet it's still easy to read and write. Personally, I like this explicit markup syntax over markdown. As the [Zen of Python](https://peps.python.org/pep-0020/) states, "Explicit is better than implicit."

**Reply by amitu** [reply]
I agree with the advantages of your syntax over markdown. I also agree that explicit grouping using `{}` and `\foo{}~bar{}` etc do solve some of the syntax issues we are facing in our new end based syntax. 

I think your `~bar` to represent "continuation"/argument to `foo` is brilliant. Also eg `\link{Aadit Codes}~href{https://aadit.codes/}` is inline and terse. It's brilliant. 

The key problem is with your syntax is "ugliness". Maybe my eyes are used to my syntax, where I find:

```ftd
-- ds.h1: hello world

This is a body of the heading

-- ds.h2: Some sub heading

sub heading content

-- ds.code: still in sub heading
lang: py

print("hello world")
```

looks "cleaner" than:

```tex
\ds.h1 {hello world}{

{This is a body of the heading}

\ds.h2 {Some sub heading}{

{sub heading content}

\ds.code {still in sub heading} ~lang{py}{print("hello world")}

}}
```

I do not like lots of `{}`. I do like `~` to continue things.

```ftd
-- foo: Hello

-- foo~something:

whatever

-- foo~something:

whatever

-- bar:
```

Here since the second and third sections tarted with `foo~`, we know they are part of `foo` still. We could have made them `~something` as well. Or try something like:

```ftd
-- foo: Hello

~~ something:

whatever

~~ something:

whatever

-- bar:
```

Do you agree @aaditmshah that the closing all the `{}` is a problem?

**Reply by amitu** [reply]
> Note that this markup format is only for content. It shouldn't be used for presentation, document structure, or application logic. For presentation we can use CSS, for document structure we can use HTML, and for application logic we can use JavaScript. 

I do not agree with this. I mean why must this not be allowed?

```tex
\boolean{open}{true}
\h1{hello}~on-click{x = !x}{
\para{
Details
}~if{$open}
}
```

We are putting some "application logic here". If you can not abstract out this logic in reusable component then this may be considered a bad idea, but if you can do:

```tex
\component{foo}{

}
```

Actually the `{}` based syntax is proving quite ugly here. Compare it with `-- ` syntax.

With inline app logic:

```ftd
-- boolean open: true
-- h1: hello
on-click: {open = !open}

-- para:
if: $open

Details
```

Vs abstracted component:

```ftd
-- component with-detail:
caption title:
boolean open:
body details:

-- h1: $title

-- para:
if: $open

$details

-- end: with-detail
```

And using it:

```ftd
-- with-details: hello

Details
```

Why must people have to learn JavaScript, React, bundler, package manager etc all to do something this simple?

**Reply by amitu** [reply]
Sure, with JavaScript you get a lot. And we are creating wasm based `ffi` so you can write stuff in JavaScript or any language that compiles to WASM so complex logic is possible. 

For creating content, you must be able to define your own custom domain specific components, eg

```ftd
-- var: HOME
default: /home/<username>

This environment variable represents the directory which...

-- var: EDITOR
default: vi

This is the editor preferred by the user..


-- component var:
caption name:
optional string default:
body description

.. definition omitted ..
-- end: var
```

Here the component `var` is content. Depending on the content you are authoring you would want to create such custom domain objects. 

Then the definition of `var` component can be moved to a utils.ftd if you want. We have achieved the separate of content and presentation, what `var` means, vs how it looks like. But without telling the author that they have to go and learn JavaScript. Only 25M programmers are there in the world. JS developer even fewer. So why must "non programmers" be denied the power to create `var` or `person` or `meeting` or `event` etc when they are talking about these things? Why must they be forced to learn JS?

**Reply by aaditmshah** [reply]
> The key problem is with your syntax is "ugliness".

You are still fixated on the syntax of the language, which is a trivial matter as compared to the semantics of the language. This is a very common problem in language design. Indeed, it's so common that it has its own name — [Wadler's law of language design](https://wiki.haskell.org/Wadler's_Law).

> In any language design, the total time spent discussing a feature in this list is proportional to two raised to the power of its position.
>
> 0. Semantics
> 1. Syntax
> 2. Lexical syntax
> 3. Lexical syntax of comments
>
> (That is, twice as much time is spent discussing syntax than semantics, twice as much time is spent discussing lexical syntax than syntax, and twice as much time is spent discussing syntax of comments than lexical syntax.)

Wadler's law is a special case of [Parkinson's law of triviality](https://en.wikipedia.org/wiki/Law_of_triviality), which states that people spend disproportionately more time on easy and minor issues like language syntax, and less time on complex and important issues like language semantics.

The syntax of a language is like the paint on a building. The semantics of a language is like the foundation and structure of a building. When you're constructing a building, the most important thing is to design its foundation and structure. The paint can be thought of afterwards. Similarly, when designing a language the most import thing to define is its semantics. The syntax can be thought of later.

This is one of the reasons why Lisp is so popular for programming language research. The syntax of Lisp is quite ugly. However, it's great for creating embedded [domain-specific languages](https://en.wikipedia.org/wiki/Domain-specific_language) because it allows language designers to focus on the semantics of the DSL instead of worrying about the syntax. The syntax is just [s-expressions](https://en.wikipedia.org/wiki/S-expression) which maps directly to the [abstract syntax tree](https://en.wikipedia.org/wiki/Abstract_syntax_tree) of the embedded DSL.

**Reply by amitu** [reply]
@aaditmshah can you should me how would you express this:

```ftd
-- component with-detail:
caption title:
boolean open:
body details:

-- h1: $title

-- para:
if: $open

$details

-- end: with-detail
```

**Reply by amitu** [reply]
We are quite happy with the semantics. Language is working fine now. We can move on to adding more complex features. In 2 yrs we have used same syntax, and now that we are nearing the launch to a larger audience we are reviewing it. 

And the whole point of this phase is syntax. If we did not have some syntax related problems we would have moved to other features.

**Reply by aaditmshah** [reply]
> > Note that this markup format is only for content. It shouldn't be used for presentation, document structure, or application logic. For presentation we can use CSS, for document structure we can use HTML, and for application logic we can use JavaScript.
> 
> I do not agree with this. I mean why must this not be allowed?

Separation of concerns. Don't pollute your content with application logic. Data files should only contain data. They shouldn't contain logic to present, structure, or process that data.

For example, suppose you want to create an article with three paragraphs. The second paragraph should be a disclosure widget in which information is visible only when the widget is toggled to an "open" state.

> Bikeshedding is a term coined as a metaphor to illuminate [Parkinson’s Law of Triviality](https://en.wikipedia.org/wiki/Parkinson%27s_Law_of_Triviality).
>
> <details>
>   <summary>Bikeshedding</summary>
>   Futile investment of time and energy in discussion of marginal technical issues.
> </details>
>
> Parkinson observed that a committee whose job is to approve plans for a [nuclear power plant](https://en.wiktionary.org/wiki/nuclear_power_plant) may spend the majority of its time on relatively unimportant but easy-to-grasp issues, such as what materials to use for the staff bikeshed, while neglecting the design of the power plant itself, which is far more important but also far more difficult to criticize constructively.

This can be done quite easily using the syntax I proposed.

```
\article{
  \paragraph{Bikeshedding is a term coined as a metaphor to illuminate \link{Parkinson’s Law of Triviality}~href{https://en.wikipedia.org/wiki/Parkinson%27s_Law_of_Triviality}.}
  \summary{Bikeshedding}~details{Futile investment of time and energy in discussion of marginal technical issues.}
  \paragraph{Parkinson observed that a committee whose job is to approve plans for a \link{nuclear power plant}~href{https://en.wiktionary.org/wiki/nuclear_power_plant} may spend the majority of its time on relatively unimportant but easy-to-grasp issues, such as what materials to use for the staff bikeshed, while neglecting the design of the power plant itself, which is far more important but also far more difficult to criticize constructively.}
}
```

As you can see, there's no application logic in the above code snippet. Everything is just content demarcated with tags. The meaning of that content, how it's supposed to be presented, and the application logic is left for the application to provide.

In our zoom meeting on Friday you mentioned that there were two target audiences you had in mind. First, [authors](https://fpm.dev/author/) who shouldn't have to learn programming to create content. Second, [builders](https://fpm.dev/themes/) who create components for authors. By separating content from application logic, authors can focus on creating content and builders can create components in a more powerful language. Doesn't have to be an existing programming language, but why reinvent the wheel?

> Why must people have to learn JavaScript, React, bundler, package manager etc all to do something this simple?

Authors don't need to learn all that. Builders might need some programming knowledge.

**Reply by amitu** [reply]
@aaditmshah I have created [this discussion where the semantics is discussed in some detail](https://github.com/FifthTry/ftd/discussions/426). I have mentioned which parts are changing:

- remove container management (goal of current exercise)
- support function definition etc (we will do this later).

**Reply by aaditmshah** [reply]
> @aaditmshah can you should me how would you express this:

Sure, consider that a builder created the following React components.

```javascript
const Disclosure = (props) => (<details><summary>{props.summary}</summary>{props.details}</details>);
const Paragraph = (props) => (<p>{props.paragraph}</p>);
const Article = (props) => (
  <>
    {
      article.children.map((child) =>
        child.$type === "paragraph"
          ? <Paragraph paragraph={child.paragraph} />
          : <Disclosure summary={child.summary} details={child.details} />
      )
    }
  </>
);
```

Now, an author who knows nothing about programming creates the following content.

```
\article{
  \paragraph{Bikeshedding is a term coined as a metaphor to illuminate \link{Parkinson’s Law of Triviality}~href{https://en.wikipedia.org/wiki/Parkinson%27s_Law_of_Triviality}.}
  \summary{Bikeshedding}~details{Futile investment of time and energy in discussion of marginal technical issues.}
  \paragraph{Parkinson observed that a committee whose job is to approve plans for a \link{nuclear power plant}~href{https://en.wiktionary.org/wiki/nuclear_power_plant} may spend the majority of its time on relatively unimportant but easy-to-grasp issues, such as what materials to use for the staff bikeshed, while neglecting the design of the power plant itself, which is far more important but also far more difficult to criticize constructively.}
}
```

Your FTD compiler does some magic, calls the builder's component, and voila!

> Bikeshedding is a term coined as a metaphor to illuminate [Parkinson’s Law of Triviality](https://en.wikipedia.org/wiki/Parkinson%27s_Law_of_Triviality).
>
> <details>
>   <summary>Bikeshedding</summary>
>   Futile investment of time and energy in discussion of marginal technical issues.
> </details>
>
> Parkinson observed that a committee whose job is to approve plans for a [nuclear power plant](https://en.wiktionary.org/wiki/nuclear_power_plant) may spend the majority of its time on relatively unimportant but easy-to-grasp issues, such as what materials to use for the staff bikeshed, while neglecting the design of the power plant itself, which is far more important but also far more difficult to criticize constructively.

The point I'm making is that.

1. The author and the builder don't need to use the same language. In fact, it's better if they use separate languages.
2. The author, being a non-programmer, shouldn't have to write application logic.
3. The builder needs to write application logic. Hence, the builder needs to know programming. However, we don't have to reinvent the wheel. We can simply use an existing language. Possibly create an embedded domain-specific language.

**Reply by amitu** [reply]
I am separating the author and builder, but not the language, which is not needed, it works as you can see. Why must people be forced to learn JS? 

Authors need not learn about how to define components, but if they want to they can with an incremental effort learn it. I anticipate most authors to define domain components, so you can define say `api` or `http-get-api` component to document your `http get api` and do not have to rely on builders. 

Every excel sheet has a type definition. We can not expect that a totally different person creates the column and totally different person works with that excel. To work with data you must define data type. Every excel user is defining data type. 

The separation you think is clearly defined does not exist. What separation Excel has? In your opinion Excel is winning or losing product? We want to make better Excel. Which can generate any UI and not just grids. And authors will use components written by builders, but not just directly, but also by creating custom components.

We can't argue on the goals, I want to you to accept the project's goal. I am not saying other goals shouldn't exist, but this is the goal of this project. To give this power to this massive class of users. Help me do that.

**Reply by amitu** [reply]
@aaditmshah do that without requiring JavaScript/JSX etc knowledge.

**Reply by amitu** [reply]
You may agree that JS is it, the world must learn JS and sure. We are here to say something better is possible. Virtually everything you can do, for content site use case (tabs, dropdown, expand collapse, tree view, sliders, presentations etc etc) you can do *today* with FTD without learning JS. Learning JS is a huge learning curve. Even 25M programmers are not all JS programmers. The syntax that you just showed, that you can use `<>` I did not know. I have 24 years of programming experience. 

If your key argument is "they must be different language", if someone creates a language that does both, gods of programming will be unhappy, then not much can be said. If you accept the goal of the project to be "lets dream of that common language that can do both" (not fully, not full replacement of JS, but full enough). Like Elm was for building web app, Elm is 80% JS, can not do so much JS can, but good enough for what Elm targeted. Similarly FTD. Elm was for web apps and FTD is for content heavy sites.

**Reply by aaditmshah** [reply]
>  So why must "non programmers" be denied the power to create `var` or `person` or `meeting` or `event` etc when they are talking about these things? Why must they be forced to learn JS?

I don't assume to know who your target audience is or how they'll interact with your software. However, from what I understand you have two target audiences. Authors, who want to create content and don't need to know programming. Builders, who want to create reusable components and need to know programming.

Now, you're asking me why deny authors the power of creating a new component by forcing them to learn JavaScript to build a new component. First, I thought authors are a different set of users than builders. I thought authors didn't need to know about programming at all. Second, why create an entirely new programming language when you can just create an embedded domain-specific language in an existing programming language? It seems that you're unnecessarily reinventing the wheel.

Now, there's nothing wrong with reinventing the wheel. However, a half-baked language will probably do you more harm than good in the long run. Creating a new general-purpose programming language is quite difficult. On the other hand, creating an embedded domain-specific language is much simpler.

**Reply by amitu** [reply]
They are different but the gap between them must be massively large is not acceptable. It must be one language and the path between author to builder must be small. Markdown/JSX etc exists if that was not the case. I am building FTD because this massive gap is not acceptable to me. And why are you calling it half baked etc? 

Please accept the goal there there are these people want to create something nice and help us, or you do not want to, do something you like. All this calling our work half baked, un-necessary, doing harm to someone etc is probably not worth your time right? I mean is this what you do for all language building repos? Go and tell them just use JS or why have you created this new half baked language etc? 

We are trying to create a language that gives power to 100s of millions of people. What are you trying to achieve?

**Reply by aaditmshah** [reply]
> @aaditmshah do that without requiring JavaScript/JSX etc knowledge.

You're asking me to create an entirely new DSL off the top of my head. I can't do that. Nobody can.

I also don't really understand the problem that you're trying to solve, how you're trying to solve that problem, or what challenges you're currently facing in implementing your solution. I'm coming to this discussion pretty much blindfolded. So, the only advice that I can give, based on my limited understanding of your product, is what I think you ought to be focusing on.

**Reply by aaditmshah** [reply]
> Please accept the goal there there are these people want to create something nice and help us, or you do not want to, do something you like.

All right. I'll see myself out then. All the best for your project.

**Reply by amitu** [reply]
Got it @aaditmshah.

The high level problem: create a language that gives 80% power of HTML/CSS/JS to non programmers. The [semantics of language](https://github.com/FifthTry/ftd/discussions/426) is mostly fine. 

The syntax is our problem.

We have a current syntax, this is quite dated:

```ftd

-- ftd.row:

-- ftd.text: hello

-- ftd.text: world

-- container: ftd.main
```

This syntax does not have a `end` clause. We do explicit [container management](http://ftd.dev/container-management/) for building arbitrary nested tree using a flat looking structure. Container management is our "invention", we are not fond of it any more. We have spent a lot of time teaching that to new comers. Currently it takes 2-3 weeks for a non programmer to learn ftd, and majority of this is in grokking container management. Also being flat means some other things are harder than need to be.

So we are considering introducing `end`:

```ftd

-- ftd.row:

-- ftd.text: hello

-- ftd.text: world

-- end: ftd.row
```

It leads to some tricky stuff, which we have discussed in [details in different discussions](https://github.com/FifthTry/ftd/discussions/388). We are also trying to achieve a. no white space sensitivity, b. more importantly, as much as possible readable code when when white space has not been used. 

b. needs some clarification: most non white space sensitive languages readability still requires indentation. We want a language that has as little indentation as possible. Goal is 100s of millions of people, and they are not going to have indentation aware editors. So this is a language goal.

We can review the different things we would want to express in the proposal we have (we have a concrete design on hand, just so you are sure). We want help with someone reviewing those.

Or proposing a new syntax. I quite liked the syntax you proposed. It had a few properties we have. Would that be overall syntax we go with? Not till we can fully do everything we can do with our current syntax using the that syntax. That is table stakes. We are not going to have two language. 

So within this set of rules, you can review the syntax we have. Arpita is pending a lot of hours in implementing that syntax. If you see the final product, and then you give me some ideas it may be too much for me to ask her to rewrite things (to be frank I have mentally resigned to the possibility that whatever we are doing next, will be revamped after I get feedback from larger audience). So if I can get some earlier feedback it would be good.

Maybe I should create a new discussion and show all sorts of things you can do in the new syntax in a cleaned up discussion instead of linking the dozen or so proposals and discussions we have currently. 

Till then we can review the [language semantics](https://github.com/FifthTry/ftd/discussions/426).
